### PR TITLE
feat(http gateway)

### PR DESCRIPTION
**Pull request by cong-or**
Initial HTTP gateway scaffold.

Future PRs which expand on this:
- extension of connection manager to include event tracking.
- example hello world kv app with defined API surface and wasm component to showcase - "how everything works in tandem"

### CODE REVIEW

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        on_http_event,
+        TargetApp::All,
+        TargetModule::All,
+    ))?;
+
+    match &receiver.recv()? {
+        HTTPEventMsg::HttpEventResponseSome(resp) => {
+            Ok(Response::new(serde_json::to_string(&resp)?.into()))
+        },
+        HTTPEventMsg::HttpEventResponseNone() => Ok(not_found()),
```
Comment (cong-or) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs:164]: @stevenj in the case where the event responds with nothing (none). What should we return to the client? Should we classify nothing returned as a 404?

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    connection_manager
+        .connection_context
+        .try_lock()
+        .map_err(|_| anyhow::anyhow!("Unable to obtain mutex lock"))?
+        .insert(
+            unique_request_id,
+            (ClientIPAddr(ip), Processed(true), LiveConnection(false)),
+        );
+
+    info!("connection manager {:?}\n", connection_manager);
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs:103]:
```suggestion
    info!("connection manager {:?}", connection_manager);
```

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+    for (header_name, header_val) in req.headers() {
+        header_map
+            .entry(header_name.to_string())
+            .or_default()
+            .push(header_val.to_str()?.to_string());
+    }
+
+    match uri.path() {
+        "/api" => {
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs:127]: I think `"api"` value should be as a const value

**Review by Mr-Leshiy — CHANGES_REQUESTED** [review]
Am I correct that mechanics of this http extensions is follows right now:
http runtime extensions receives request -> for the received request transformed to hermes event -> hermes http event is sent to event queue -> hermes event executes inside wasm -> original request sender receives response after the successful event execution as a part of the event processing function.

What happens if event will be not executed inside wasm (because event queue does not provide such guaranties). Does the user will receive some response at this case ?

**Comment by cong-or** [comment]
> Am I correct that mechanics of this http extensions is follows right now: http runtime extensions receives request -> for the received request transformed to hermes event -> hermes http event is sent to event queue -> hermes event executes inside wasm -> original request sender receives response after the successful event execution as a part of the event processing function.
> 
> What happens if event will be not executed inside wasm (because event queue does not provide such guaranties). Does the user will receive some response at this case ?

yes correct. But the next step is to add event tracking and management - the connection manager is a start for this. We will think about it in more detail then. At the moment we just respond [like this ](https://github.com/input-output-hk/hermes/blob/feature-http-gateway/hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs#L160) - the event either returns some or None and the http response mirrors this back to the client.

File: hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs (hunk 1 above)
Comment (stevenj) [reply hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs:171]: We should not return anyting, UNTIL all wasm modules have finished.  At which time if none of them supplied a response we return 404.

**Comment by cong-or** [comment]
re-opening with dependency tracking changes.

**Pull request closed without merging**